### ISSUE TITLE
Timeout with wkc = EC_NOCRAME. Falsely

### ISSUE THREAD

**Issue by mwh-newtec** [OpenEtherCATsociety/SOEM#595]
Hi,

In ecx_inframe() a thread checks the rxbuf at it's index for a recieved message, and in case nothing has been received yet, it proceeds to take the rx mutex and call recv() waiting for a message.

Please look a the following case:
Thread A has higher priority than B.
Thread B has lower priority than A.

So thread B initiates a mailbox communication (slow), sends, checks rxbuf, locks rx mutex and calls recv() waiting for the message to return or timing out (timeouts align to system timer granularity on linux).
Meanwhile thread A initiates a process data send()/receive(). It sends, checks rxbuf, and waits to lock rx mutex.
Thread B gets elevated to the priority of B, but we're still waiting to receive a message so can't proceed.
Now the response of thread A gets received. Thread B puts it in rxbuf and unlocks mutex.
Straight away thread A is scheduled in and starts a recv() call, not knowing it's messages has already been received. 
Thread B hasn't timed out, so it checks rxbuf, and waits to lock rx mutex.
Thread A receives thread B's message, puts it in rxbuf and unlocks mutex.

If thread A has timed out now, we return EC_NOFRAME, whilst having received the message.
Thread B could have the same fate depending on the work load of thread A.

Even worse if thread A sends a mailbox communication and thread B sends/receives process data.
The result would be thread A holding the mutex for multiple milliseconds (recv() timeout), unless receiving B's second process data.
That would then lead to thread B holding the mutex for multiple milliseconds (recv() timeout).
^^Has been observed leading.

This can be circumvented by checking rxbuf again after getting the rxbuf and skipping recv() in case some one else has received our message.

Best regards Mathias

**Comment by ArthurKetels** [comment]
Yes, this is a known behaviour. And as you posted in #451 the cause is the timeout behaviour of the Linux kernel. In older kernel versions this used to work properly. In newer versions not so much. Solving the proper non-blocking behaviour for recv() would remove this issue too.

On the other hand, adding the extra check after obtaining the mutex is simple and would improve the timing.
